org.jacoco.core/src/org/jacoco/core/instr/Instrumenter.java

```diff
@@ -200,6 +203,12 @@ public int instrumentAll(final InputStream input,
 
 	private int instrumentZip(final InputStream input,
 			final OutputStream output, final String name) throws IOException {
+		if (generators.getLast() instanceof
+				OfflineInstrumentationCompanionAccessGenerator) {
+			generators
+					.addLast(new
+							OfflineInstrumentationCompanionAccessGenerator());
```

Comment:
Well, yes, but I think it's obvious that setRemoveSignatures mutates the instance and should be used to initialize the instance, but I think it's less obvious that instrumentAll is thread-unsafe.  Anyway, since the generator is already thread-unsafe, I think it's fine if Instrumenter becomes thread-unsafe when using that generator.